--- /nix/store/wc1xz4in479bgzdiqh9lb78h7j8gqjkf-source/JoyShockMapper/src/SDLWrapper.cpp	1969-12-31 19:00:01.000000000 -0500
+++ /tmp/pi-github-repos/evan1mclean/JSM_custom_curve/JoyShockMapper/src/SDLWrapper.cpp	2026-06-23 15:28:22.323955496 -0400
@@ -45,6 +45,8 @@
 	ControllerDevice(int id)
 	  : _has_accel(false)
 	  , _has_gyro(false)
+	  , _vendorId(JS_VENDOR_UNKNOWN)
+	  , _productId(JS_PRODUCT_UNKNOWN)
 	{
 		_prevTouchState.t0Down = false;
 		_prevTouchState.t1Down = false;
@@ -74,8 +76,93 @@
 						SDL_SetGamepadSensorEnabled(_sdlController, SDL_SENSOR_ACCEL, true);
 					}
 
-					int vid = SDL_GetGamepadVendor(_sdlController);
-					int pid = SDL_GetGamepadProduct(_sdlController);
+					_vendorId = SDL_GetGamepadVendor(_sdlController);
+					_productId = SDL_GetGamepadProduct(_sdlController);
+					_guid = SDL_GetJoystickGUID(SDL_GetGamepadJoystick(_sdlController));
+					_ctrlr_type = JS_TYPE_UNKNOWN;
+
+					switch (_vendorId)
+					{
+					case JS_VENDOR_8BITDO:
+						switch (_productId)
+						{
+						case JS_PRODUCT_8BITDO_SF30_PRO:
+							_ctrlr_type = JS_TYPE_8BITDO_SF30_PRO;
+							break;
+						case JS_PRODUCT_8BITDO_SF30_PRO_BT:
+							_ctrlr_type = JS_TYPE_8BITDO_SF30_PRO_BT;
+							break;
+						case JS_PRODUCT_8BITDO_SN30_PRO:
+							_ctrlr_type = JS_TYPE_8BITDO_SN30_PRO;
+							break;
+						case JS_PRODUCT_8BITDO_SN30_PRO_BT:
+							_ctrlr_type = JS_TYPE_8BITDO_SN30_PRO_BT;
+							break;
+						case JS_PRODUCT_8BITDO_PRO_2:
+							_ctrlr_type = JS_TYPE_8BITDO_PRO_2;
+							break;
+						case JS_PRODUCT_8BITDO_PRO_2_BT:
+							_ctrlr_type = JS_TYPE_8BITDO_PRO_2_BT;
+							break;
+						case JS_PRODUCT_8BITDO_PRO_3:
+							_ctrlr_type = JS_TYPE_8BITDO_PRO_3;
+							break;
+						case JS_PRODUCT_8BITDO_ULTIMATE2_WIRELESS:
+							_ctrlr_type = JS_TYPE_8BITDO_ULTIMATE2_WIRELESS;
+							break;
+						}
+						break;
+					case JS_VENDOR_HORI:
+						if (_productId == JS_PRODUCT_HORI_STEAM_CONTROLLER ||
+							_productId == JS_PRODUCT_HORI_STEAM_CONTROLLER_BT)
+						{
+							_ctrlr_type = JS_TYPE_HORI_STEAM;
+						}
+						break;
+					case JS_VENDOR_FLYDIGI_V1:
+					case JS_VENDOR_FLYDIGI_V2:
+						if ((_vendorId == JS_VENDOR_FLYDIGI_V1 &&
+							 _productId == JS_PRODUCT_FLYDIGI_V1_GAMEPAD) ||
+							(_vendorId == JS_VENDOR_FLYDIGI_V2 &&
+							 (_productId == JS_PRODUCT_FLYDIGI_V2_APEX ||
+							  _productId == JS_PRODUCT_FLYDIGI_V2_VADER)))
+						{
+							switch (_guid.data[15])
+							{
+							case JS_FLYDIGI_APEX5:
+								_ctrlr_type = JS_TYPE_FLYDIGI_APEX5;
+								break;
+							case JS_FLYDIGI_VADER3_PRO:
+								_ctrlr_type = JS_TYPE_FLYDIGI_VADER3_PRO;
+								break;
+							case JS_FLYDIGI_VADER4_PRO:
+								_ctrlr_type = JS_TYPE_FLYDIGI_VADER4_PRO;
+								break;
+							case JS_FLYDIGI_VADER5_PRO:
+								_ctrlr_type = JS_TYPE_FLYDIGI_VADER5_PRO;
+								break;
+							}
+						}
+						break;
+					case JS_VENDOR_GAMESIR:
+						if (_productId == JS_PRODUCT_GAMESIR_GAMEPAD_G7_PRO_8K &&
+							_guid.data[0] == JS_HARDWARE_BUS_USB) // No extended features over Bluetooth
+						{
+							_ctrlr_type = JS_TYPE_G7_PRO_8K;
+						}
+						break;
+					case JS_VENDOR_NINTENDO:
+						if (_productId == JS_PRODUCT_NINTENDO_SWITCH2_PRO)
+						{
+							_ctrlr_type = JS_TYPE_SWITCH2_PRO_CONTROLLER;
+						}
+						break;
+					}
+
+					if (_ctrlr_type != JS_TYPE_UNKNOWN)
+					{
+						continue;
+					}
 
 					auto sdl_ctrlr_type = SDL_GetGamepadType(_sdlController);
 					switch (sdl_ctrlr_type)
@@ -100,33 +187,28 @@
 						break;
 					case SDL_GamepadType::SDL_GAMEPAD_TYPE_XBOXONE:
 						_ctrlr_type = JS_TYPE_XBOXONE;
-						if (vid == 0x0e6f) // PDP Vendor ID
+						switch (_vendorId)
 						{
+						case JS_VENDOR_PDP:
+						case JS_VENDOR_POWERA:
 							_ctrlr_type = JS_TYPE_XBOX_SERIES;
-						}
-						if (vid == 0x24c6) // PowerA
-						{
-							_ctrlr_type = JS_TYPE_XBOX_SERIES;
-						}
-						if (vid == 0x045e) // Microsoft Vendor ID
-						{
-							switch (pid)
+							break;
+						case JS_VENDOR_MICROSOFT:
+							switch (_productId)
 							{
-							case(0x02e3): // Xbox Elite Series 1
-								// Intentional fall through to the next case
-							case(0x0b05): //Xbox Elite Series 2 - Bluetooth
-								// Intentional fall through to the next case
-							case(0x0b00): //Xbox Elite Series 2
-							case (0x02ff): //XBOXGIP driver software PID - not sure what this is, might be from Valve's driver for using Elite paddles
-								// in any case, this is what my ELite Series 2 is showing as currently, so adding it here for now    
+							case JS_PRODUCT_XBOX_ONE_ELITE_SERIES_1:
+							case JS_PRODUCT_XBOX_ONE_ELITE_SERIES_2:
+							case JS_PRODUCT_XBOX_ONE_ELITE_SERIES_2_BLUETOOTH:
+							case JS_PRODUCT_XBOX_ONE_ELITE_SERIES_2_BLE:
+							case JS_PRODUCT_XBOX_ONE_XBOXGIP_CONTROLLER:
 								_ctrlr_type = JS_TYPE_XBOXONE_ELITE;
 								break;
-							case(0x0b12): //Xbox Series controller
-								// Intentional fall through to the next case
-							case(0x0b13): // Xbox Series controller - bluetooth
+							case JS_PRODUCT_XBOX_SERIES_X:
+							case JS_PRODUCT_XBOX_SERIES_X_BLE:
 								_ctrlr_type = JS_TYPE_XBOX_SERIES;
 								break;
 							}
+							break;
 						}
 						break;
 					}
@@ -221,7 +303,10 @@
 	bool _has_gyro;
 	bool _has_accel;
 	int _split_type = JS_SPLIT_TYPE_FULL;
-	int _ctrlr_type = 0;
+	int _ctrlr_type = JS_TYPE_UNKNOWN;
+	int _vendorId = JS_VENDOR_UNKNOWN;
+	int _productId = JS_PRODUCT_UNKNOWN;
+	SDL_GUID _guid;
 	uint16_t _small_rumble = 0;
 	uint16_t _big_rumble = 0;
 	AdaptiveTriggerSetting _leftTriggerEffect;
@@ -233,6 +318,157 @@
 
 struct SdlInstance : public JslWrapper
 {
+private:
+#ifdef _WIN32
+	// Make Windows 11 honor timer resolution when window is minimized or
+	// another application is fullscreen. EcoQoS is also disabled.
+	// https://learn.microsoft.com/en-us/windows/win32/api/timeapi/nf-timeapi-timebeginperiod
+	// https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessinformation
+	void DisableProcessPowerThrottling()
+	{
+		PROCESS_POWER_THROTTLING_STATE state;
+		memset(&state, 0, sizeof(state));
+		state.Version = PROCESS_POWER_THROTTLING_CURRENT_VERSION;
+		state.ControlMask = PROCESS_POWER_THROTTLING_EXECUTION_SPEED
+		                    | PROCESS_POWER_THROTTLING_IGNORE_TIMER_RESOLUTION;
+		state.StateMask = 0;
+		SetProcessInformation(GetCurrentProcess(), ProcessPowerThrottling,
+		                      &state, sizeof(state));
+	}
+
+	// Raise process priority, but not too high.
+	// https://learn.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities
+	void RaiseProcessPriority()
+	{
+		SetPriorityClass(GetCurrentProcess(), ABOVE_NORMAL_PRIORITY_CLASS);
+	}
+
+	// Raise thread priority, but not too high.
+	// https://learn.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities
+	void RaiseThreadPriority()
+	{
+		SetThreadPriority(GetCurrentThread(), THREAD_PRIORITY_HIGHEST);
+	}
+
+	typedef long NTSTATUS;
+	typedef NTSTATUS (NTAPI *PZEQTR)(PULONG MinRes, PULONG MaxRes, PULONG CurrentRes);
+	typedef NTSTATUS (NTAPI *PZESTR)(ULONG DesiredRes, BOOLEAN SetRes, PULONG CurrentRes);
+	PZEQTR ZwQueryTimerResolution = nullptr;
+	PZESTR ZwSetTimerResolution = nullptr;
+	ULONG win_timer_res = 0;
+	uint64_t timer_res_ns = 0;
+
+	// Reduce system clock interrupt interval to 0.5 ms. Windows default is
+	// 15.625 ms (1000/64). SDL default is 1 ms but it uses timeBeginPeriod.
+	void SetMaxTimerResolution()
+	{
+		ZwQueryTimerResolution = (PZEQTR)GetProcAddress(
+		    GetModuleHandle(TEXT("ntdll.dll")), "ZwQueryTimerResolution");
+		ZwSetTimerResolution = (PZESTR)GetProcAddress(
+		    GetModuleHandle(TEXT("ntdll.dll")), "ZwSetTimerResolution");
+		ULONG min_res = 0, max_res = 0, cur_res = 0;
+		bool result = false;
+
+		if (ZwQueryTimerResolution != nullptr
+		    && ZwQueryTimerResolution(&min_res, &max_res, &cur_res) == 0)
+		{
+			if (ZwSetTimerResolution != nullptr
+			    && ZwSetTimerResolution(max_res, TRUE, &cur_res) == 0)
+			{
+				win_timer_res = cur_res;
+				timer_res_ns = win_timer_res * 100; // 100-ns to ns.
+				result = true;
+			}
+		}
+
+		if (!result)
+		{
+			// Don't call again.
+			ZwQueryTimerResolution = nullptr;
+			ZwSetTimerResolution = nullptr;
+
+			// Use safe defaults.
+			win_timer_res = 10000;  // 1 ms in 100-ns units.
+			timer_res_ns = 1000000; // 1 ms.
+		}
+	}
+
+	// Call before using a waitable timer to ensure resolution is still correct.
+	void ReapplyMaxTimerRes()
+	{
+		if (ZwSetTimerResolution != nullptr)
+		{
+			ULONG cur_res = 0;
+			ZwSetTimerResolution(win_timer_res, TRUE, &cur_res);
+		}
+	}
+
+	uint64_t next_poll_time = 0;
+
+	void InitPollingTimer()
+	{
+		next_poll_time = SDL_GetTicksNS();
+	}
+
+	void PollingTimer(uint64_t interval_ms)
+	{
+		const uint64_t interval_ns = interval_ms * 1000000ULL;
+		next_poll_time += interval_ns;
+
+		uint64_t now = SDL_GetTicksNS();
+		if (now < next_poll_time)
+		{
+			// Sleep when delay is longer than timer resolution.
+			const uint64_t delay_thresh_ns = now + timer_res_ns;
+			if (delay_thresh_ns < next_poll_time)
+			{
+				// Leave a gap equal to the timer resolution.
+				const uint64_t delay_ns = next_poll_time - delay_thresh_ns;
+				ReapplyMaxTimerRes();
+				SDL_DelayNS(delay_ns);
+				now = SDL_GetTicksNS();
+			}
+
+			// Busy-wait for the remaining time.
+			while (now < next_poll_time)
+			{
+				SDL_CPUPauseInstruction();
+				now = SDL_GetTicksNS();
+			}
+		}
+		else
+		{
+			// Fell behind.
+			next_poll_time = now;
+		}
+	}
+#else
+	void DisableProcessPowerThrottling()
+	{
+	}
+
+	void RaiseProcessPriority()
+	{
+	}
+
+	void RaiseThreadPriority()
+	{
+	}
+
+	void SetMaxTimerResolution()
+	{
+	}
+
+	void InitPollingTimer()
+	{
+	}
+
+	void PollingTimer(uint64_t interval_ms)
+	{
+		SDL_DelayNS(interval_ms * 1000000ULL);
+	}
+#endif // _WIN32
+
 public:
 	SdlInstance()
 	{
@@ -249,7 +485,9 @@
 		SDL_SetHint(SDL_HINT_JOYSTICK_HIDAPI_SWITCH_HOME_LED, "0");
 		SDL_SetHint(SDL_HINT_JOYSTICK_ENHANCED_REPORTS, "1");
 		SDL_SetHint(SDL_HINT_JOYSTICK_THREAD, "1");
+		SDL_SetHintWithPriority(SDL_HINT_TIMER_RESOLUTION, "1", SDL_HINT_OVERRIDE);
 		SDL_Init(SDL_INIT_GAMEPAD);
+		SetMaxTimerResolution();
 	}
 
 	virtual ~SdlInstance()
@@ -259,10 +497,13 @@
 
 	int pollDevices()
 	{
+		RaiseThreadPriority();
+		InitPollingTimer();
+
 		while (keep_polling)
 		{
 			auto tick_time = SettingsManager::get<float>(SettingID::TICK_TIME)->value();
-			SDL_Delay(Uint32(tick_time));
+			PollingTimer(uint64_t(tick_time));
 
 			lock_guard guard(controller_lock);
 			SDL_UpdateGamepads();
@@ -299,6 +540,9 @@
 
 	int ConnectDevices() override
 	{
+		DisableProcessPowerThrottling();
+		RaiseProcessPriority();
+
 		bool isFalse = false;
 		if (keep_polling.compare_exchange_strong(isFalse, true))
 		{
@@ -392,7 +636,7 @@
 		{
 			array<float, 3> accel;
 			SDL_GetGamepadSensorData(_controllerMap[deviceId]->_sdlController, SDL_SENSOR_ACCEL, &accel[0], 3);
-			static constexpr float toGs = 1.f / 9.8f;
+			static constexpr float toGs = 1.f / SDL_STANDARD_GRAVITY;
 			imuState.accelX = accel[0] * toGs;
 			imuState.accelY = accel[1] * toGs;
 			imuState.accelZ = accel[2] * toGs;
@@ -409,6 +653,14 @@
 	{
 		TOUCH_STATE state;
 		memset(&state, 0, sizeof(TOUCH_STATE));
+
+		if (_controllerMap[deviceId] == nullptr ||
+			_controllerMap[deviceId]->_sdlController == nullptr ||
+			SDL_GetNumGamepadTouchpads(_controllerMap[deviceId]->_sdlController) <= 0)
+		{
+			return state;
+		}
+
 		if (!SDL_GetGamepadTouchpadFinger(_controllerMap[deviceId]->_sdlController, 0, 0, &state.t0Down, &state.t0X, &state.t0Y, nullptr) || 
 			!SDL_GetGamepadTouchpadFinger(_controllerMap[deviceId]->_sdlController, 0, 1, &state.t1Down, &state.t1X, &state.t1Y, nullptr))
 		{
@@ -441,9 +693,9 @@
 		return false;
 	}
 
-	int GetButtons(int deviceId) override
+	uint64_t GetButtons(int deviceId) override
 	{
-		static const map<int, int> sdl2jsl = {
+		static const map<int, uint64_t> sdl2jsl = {
 			{ SDL_GAMEPAD_BUTTON_SOUTH, JSOFFSET_S },
 			{ SDL_GAMEPAD_BUTTON_EAST, JSOFFSET_E },
 			{ SDL_GAMEPAD_BUTTON_WEST, JSOFFSET_W },
@@ -461,46 +713,114 @@
 			{ SDL_GAMEPAD_BUTTON_DPAD_RIGHT, JSOFFSET_RIGHT }
 		};
 
-		int buttons = 0;
+		uint64_t buttons = 0;
 		for (auto pair : sdl2jsl)
 		{
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GamepadButton(pair.first)) ? 1 << pair.second : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GamepadButton(pair.first)) ? 1ULL << pair.second : 0;
 
 		}
 		switch (_controllerMap[deviceId]->_ctrlr_type)
 		{
 		case JS_TYPE_JOYCON_LEFT:
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1 << JSOFFSET_CAPTURE : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1 << JSOFFSET_SL : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1 << JSOFFSET_SR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1ULL << JSOFFSET_CAPTURE : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1ULL << JSOFFSET_SR : 0;
 			break;
 		case JS_TYPE_JOYCON_RIGHT:
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1 << JSOFFSET_SL : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1 << JSOFFSET_SR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1ULL << JSOFFSET_SL : 0;
+			break;
+		case JS_TYPE_PRO_CONTROLLER:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1ULL << JSOFFSET_CAPTURE : 0;
+			break;
+		case JS_TYPE_SWITCH2_PRO_CONTROLLER:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1ULL << JSOFFSET_CAPTURE : 0;    // Capture button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0; // GR back button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;  // GL back button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC2) ? 1ULL << JSOFFSET_MISC1 : 0;      // C button
 			break;
 		case JS_TYPE_DS:
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1 << JSOFFSET_MIC : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1 << JSOFFSET_SR : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1 << JSOFFSET_SL : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1 << JSOFFSET_FNR : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1 << JSOFFSET_FNL : 0;
-			// Intentional fall through to the next case
-		case JS_TYPE_DS4:
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_TOUCHPAD) ? 1 << JSOFFSET_CAPTURE : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1 << JSOFFSET_SL : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1 << JSOFFSET_SR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1ULL << JSOFFSET_MIC : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_TOUCHPAD) ? 1ULL << JSOFFSET_CAPTURE : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1ULL << JSOFFSET_FNR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1ULL << JSOFFSET_FNL : 0;
 			break;
-		case JS_TYPE_PRO_CONTROLLER:
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1 << JSOFFSET_CAPTURE : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1 << JSOFFSET_SR : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1 << JSOFFSET_SL : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1 << JSOFFSET_FNR : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1 << JSOFFSET_FNL : 0;
+		case JS_TYPE_DS4:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_TOUCHPAD) ? 1ULL << JSOFFSET_CAPTURE : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1ULL << JSOFFSET_SR : 0;
+			break;
+		case JS_TYPE_HORI_STEAM:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0;  // R4 back button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;   // L4 back button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1ULL << JSOFFSET_FNR : 0; // M2 button below right stick
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1ULL << JSOFFSET_FNL : 0;  // M1 button below left stick
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC2) ? 1ULL << JSOFFSET_MISC1 : 0;       // QAM button ("..." button)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC3) ? 1ULL << JSOFFSET_LTOUCH : 0;      // Left stick capacitive touch
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC4) ? 1ULL << JSOFFSET_RTOUCH : 0;      // Right stick capacitive touch
+			break;
+		case JS_TYPE_G7_PRO_8K:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1ULL << JSOFFSET_CAPTURE : 0;     // Share button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0;  // R4 back button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;   // L4 back button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC2) ? 1ULL << JSOFFSET_LMINI : 0;       // L5 mini shoulder button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC3) ? 1ULL << JSOFFSET_RMINI : 0;       // R5 mini shoulder button
+			break;
+		// 8BitDo controllers with gyro and no additional buttons.
+		case JS_TYPE_8BITDO_SF30_PRO:
+		case JS_TYPE_8BITDO_SF30_PRO_BT:
+		case JS_TYPE_8BITDO_SN30_PRO:
+		case JS_TYPE_8BITDO_SN30_PRO_BT:
+			break;
+		// 8BitDo controllers with gyro and two additional buttons.
+		case JS_TYPE_8BITDO_PRO_2:
+		case JS_TYPE_8BITDO_PRO_2_BT:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0; // P1 back button (right)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;  // P2 back button (left)
+			break;
+		// 8BitDo controllers with gyro and four additional buttons.
+		case JS_TYPE_8BITDO_PRO_3:
+		case JS_TYPE_8BITDO_ULTIMATE2_WIRELESS:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_RMINI : 0; // R4 mini shoulder button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_LMINI : 0;  // L4 mini shoulder button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1ULL << JSOFFSET_SR : 0;    // PR back button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1ULL << JSOFFSET_SL : 0;     // PL back button
+			break;
+		case JS_TYPE_FLYDIGI_APEX5:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0;  // M1 back button (top right)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;   // M2 back button (top left)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1ULL << JSOFFSET_FNR : 0; // M3 back button (bottom left)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1ULL << JSOFFSET_FNL : 0;  // M4 back button (bottom right)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC2) ? 1ULL << JSOFFSET_LMINI : 0;       // LM mini shoulder button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC3) ? 1ULL << JSOFFSET_RMINI : 0;       // RM mini shoulder button
+			break;
+		case JS_TYPE_FLYDIGI_VADER5_PRO:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC4) ? 1ULL << JSOFFSET_LMINI : 0;       // LM mini shoulder button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC5) ? 1ULL << JSOFFSET_RMINI : 0;       // RM mini shoulder button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC6) ? 1ULL << JSOFFSET_MISC3 : 0;       // Circle button below right stick
+			// Fall through.
+		case JS_TYPE_FLYDIGI_VADER4_PRO:
+		case JS_TYPE_FLYDIGI_VADER3_PRO:
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0;  // M1 back button (top right)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;   // M2 back button (top left)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1ULL << JSOFFSET_FNR : 0; // M3 back button (bottom left)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1ULL << JSOFFSET_FNL : 0;  // M4 back button (bottom right)
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC2) ? 1ULL << JSOFFSET_MISC1 : 0;       // C face button
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC3) ? 1ULL << JSOFFSET_MISC2 : 0;       // Z face button
 			break;
 		default:
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1 << JSOFFSET_CAPTURE : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1 << JSOFFSET_FNL : 0;
-			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1 << JSOFFSET_FNR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC1) ? 1ULL << JSOFFSET_MISC1 : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE1) ? 1ULL << JSOFFSET_SR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE1) ? 1ULL << JSOFFSET_SL : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_RIGHT_PADDLE2) ? 1ULL << JSOFFSET_FNR : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_LEFT_PADDLE2) ? 1ULL << JSOFFSET_FNL : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC2) ? 1ULL << JSOFFSET_MISC2 : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC3) ? 1ULL << JSOFFSET_MISC3 : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC4) ? 1ULL << JSOFFSET_MISC4 : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC5) ? 1ULL << JSOFFSET_MISC5 : 0;
+			buttons |= SDL_GetGamepadButton(_controllerMap[deviceId]->_sdlController, SDL_GAMEPAD_BUTTON_MISC6) ? 1ULL << JSOFFSET_MISC6 : 0;
 			break;
 		}
 		return buttons;
@@ -627,6 +947,16 @@
 		return float();
 	}
 
+	float GetSampleRateHz(int deviceId) override
+	{
+		return 0.0f;
+	}
+
+	float GetTimeSinceLastUpdate(int deviceId) override
+	{
+		return 0.0f;
+	}
+
 	void ResetContinuousCalibration(int deviceId) override
 	{
 	}
@@ -669,6 +999,16 @@
 		return _controllerMap[deviceId]->_split_type;
 	}
 
+	int GetControllerVendor(int deviceId) override
+	{
+		return _controllerMap[deviceId]->_vendorId;
+	}
+
+	int GetControllerProduct(int deviceId) override
+	{
+		return _controllerMap[deviceId]->_productId;
+	}
+
 	int GetControllerColour(int deviceId) override
 	{
 		return int();
